crates/runtime-config/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+    pub runtime_config: T,
+    /// The resolver used to resolve key-value stores from runtime configuration.
+    pub key_value_resolver: key_value::RuntimeConfigResolver,
+}
+
+impl<T> ResolvedRuntimeConfig<T>
+where
+    T: for<'a> TryFrom<TomlRuntimeConfigSource<'a>>,
+    for<'a> <T as TryFrom<TomlRuntimeConfigSource<'a>>>::Error: Into<anyhow::Error>,
+{
+    /// Creates a new resolved runtime configuration from a runtime config source TOML file.
+    pub fn from_file(runtime_config_path: &Path, state_dir: Option<&str>) -> anyhow::Result<Self> {
+        let key_value_resolver = key_value_resolver(PathBuf::from(
+            state_dir.unwrap_or_else(|| DEFAULT_STATE_DIR.into()),
+        ));
+
+        let file = std::fs::read_to_string(runtime_config_path).with_context(|| {
+            format!(
+                "failed to read runtime config file '{}'",
+                runtime_config_path.display()
+            )
+        })?;
+        let toml = toml::from_str(&file).with_context(|| {
+            format!(
+                "failed to parse runtime config file '{}' as toml",
+                runtime_config_path.display()
+            )
+        })?;
+        let runtime_config: T = TomlRuntimeConfigSource::new(&toml, &key_value_resolver)
+            .try_into()
+            .map_err(Into::into)?;
+
+        Ok(Self {
+            runtime_config,
+            key_value_resolver,
+        })
+    }
+}
+
+/// The TOML based runtime configuration source Spin CLI.
+pub struct TomlRuntimeConfigSource<'a> {
+    table: TomlKeyTracker<'a>,
+    key_value: &'a key_value::RuntimeConfigResolver,
+}
+
+impl<'a> TomlRuntimeConfigSource<'a> {
+    pub fn new(table: &'a toml::Table, key_value: &'a key_value::RuntimeConfigResolver) -> Self {
+        Self {
+            table: TomlKeyTracker::new(table),
+            key_value,
+        }
+    }
+}
+
+impl FactorRuntimeConfigSource<WasiFactor> for TomlRuntimeConfigSource<'_> {
+    fn get_runtime_config(&mut self) -> anyhow::Result<Option<()>> {
+        Ok(None)
+    }
+}
+
+impl FactorRuntimeConfigSource<KeyValueFactor> for TomlRuntimeConfigSource<'_> {
+    fn get_runtime_config(
+        &mut self,
+    ) -> anyhow::Result<Option<spin_factor_key_value::RuntimeConfig>> {
+        self.key_value.resolve_from_toml(Some(self.table.as_ref()))
+    }
+}
+
+impl RuntimeConfigSourceFinalizer for TomlRuntimeConfigSource<'_> {
+    fn finalize(&mut self) -> anyhow::Result<()> {
+        Ok(self.table.validate_all_keys_used()?)
+    }
+}
+
+const DEFAULT_SPIN_STORE_FILENAME: &str = "sqlite_key_value.db";
+
+/// The key-value runtime configuration resolver used by the trigger.
+///
+/// Takes a base path for the local store.
+pub fn key_value_resolver(local_store_base_path: PathBuf) -> key_value::RuntimeConfigResolver {
+    let mut key_value = key_value::RuntimeConfigResolver::new();
+
+    // Register the supported store types.
+    // Unwraps are safe because the store types are known to not overlap.
```

Comment:
micro-nit: avoid the "s" word...

```suggestion
    // Unwraps are fine because the store types are known to not overlap.
```

Reply:
👍 I'll change this in the next PR to avoid having to run CI again.